### PR TITLE
Migrations functionality

### PR DESCRIPTION
**Pull request by adiweiss**
issue: https://github.com/wework/grabbit/issues/42

### CODE REVIEW

File: gbus/abstractions.go [hunk 1]
```diff
@@ -205,6 +205,7 @@ type TxProvider interface {
 	New() (*sql.Tx, error)
 	Dispose()
 	Ping(timeoutInSeconds time.Duration) bool
+	GetDb() *sql.DB
```
Comment (rhinof) [inline gbus/abstractions.go]: I think we can remove the GetDb function and pass the migration function an instance of the db struct coming from Provider.Database

File: gbus/builder/builder.go [hunk 2]
```diff
@@ -80,6 +80,7 @@ func (builder *defaultBuilder) Build(svcName string) gbus.Bus {
 				panic(err)
 			}
 			gb.TxProvider = mysqltx
+			mysql.EnsureSchema(mysqltx.GetDb(), gb.SvcName)
```
Comment (rhinof) [inline gbus/builder/builder.go]: pass in mysqltx.Database pointer instead of mysqltx.GetDb()

File: gbus/tx/provider.go [hunk 3]
```diff
@@ -38,3 +38,8 @@ func (provider *Provider) Ping(timeoutInSeconds time.Duration) bool {
 
 	return hasPing
 }
+
+func (provider *Provider) GetDb() *sql.DB {
```
Comment (rhinof) [inline gbus/tx/provider.go]: Not needed and can be removed

File: gbus/tx/mysql/migrations.go [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+		Func: func(tx *sql.Tx) error {
+			if _, err := tx.Exec(createTableQuery); err != nil {
+				return err
+			}
+			return nil
+		},
+	}
+}
+
+func EnsureSchema(db *sql.DB, svcName string) {
```
Comment (rhinof) [inline gbus/tx/mysql/migrations.go]: please add at least some basic tests

Comment (rhinof) [reply gbus/tx/mysql/migrations.go]: after thinking about it some more there is no need as this code gets executed and covered in all the existing tests

File: gbus/tx/mysql/migrations.go [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+import (
+	"database/sql"
+
+	"fmt"
+	"github.com/lopezator/migrator"
+	"github.com/wework/grabbit/gbus/tx"
+)
+
+func SagaStoreTableMigration(svcName string) *migrator.Migration {
```
Comment (rhinof) [inline gbus/tx/mysql/migrations.go]: Please add a comment for exported type describing what it does.

File: gbus/tx/mysql/migrations.go [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+		Func: func(tx *sql.Tx) error {
+			if _, err := tx.Exec(createTableQuery); err != nil {
+				return err
+			}
+			return nil
+		},
+	}
+}
+
+func OutboxMigrations(svcName string) *migrator.Migration {
```
Comment (rhinof) [inline gbus/tx/mysql/migrations.go]: Please add a comment for exported type describing what it does.

File: gbus/tx/mysql/migrations.go [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+		Func: func(tx *sql.Tx) error {
+			if _, err := tx.Exec(query); err != nil {
+				return err
+			}
+			return nil
+		},
+	}
+}
+
+func TimoutTableMigration(svcName string) *migrator.Migration {
```
Comment (rhinof) [inline gbus/tx/mysql/migrations.go]: Please add a comment for exported type describing what it does.

File: gbus/tx/mysql/migrations.go [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+		Func: func(tx *sql.Tx) error {
+			if _, err := tx.Exec(createTableQuery); err != nil {
+				return err
+			}
+			return nil
+		},
+	}
+}
+
+func EnsureSchema(db *sql.DB, svcName string) {
```
Comment (rhinof) [inline gbus/tx/mysql/migrations.go]: Please add a comment for exported type describing what it does.

File: go.sum [hunk 9]
```diff
@@ -9,6 +9,12 @@ github.com/Shopify/sarama v1.23.0 h1:slvlbm7bxyp7sKQbUwha5BQdZTqurhRoI+zbKorVigQ
 github.com/Shopify/sarama v1.23.0/go.mod h1:XLH1GYJnLVE0XCr6KdJGVJRTwY30moWNJ4sERjXX6fs=
 github.com/Shopify/toxiproxy v2.1.4+incompatible h1:TKdv8HiTLgE5wdJuEML90aBgNWsokNbMijUGhmcoBJc=
 github.com/Shopify/toxiproxy v2.1.4+incompatible/go.mod h1:OXgGpZ6Cli1/URJOF1DMxUHB2q5Ap20/P/eIdh4G0pI=
+github.com/adiweiss/migrator v0.1.1 h1:unnb8Z83UIOrjRPI/A8jS67EGQs3W8V0OtnzEJOhIuY=
```
Comment (rhinof) [inline go.sum]: remove references to github.com/adiweiss/migrator

File: go.sum [hunk 10]
```diff
@@ -9,6 +9,12 @@ github.com/Shopify/sarama v1.23.0 h1:slvlbm7bxyp7sKQbUwha5BQdZTqurhRoI+zbKorVigQ
...
+github.com/adiweiss/migrator v0.1.1/go.mod h1:bpVAVPkWSvTw8ya2Pk7E/KiNAyDWNImgivQY79o8/8I=
```
Comment (rhinof) [inline go.sum]: remove references to github.com/adiweiss/migrator

File: go.sum [hunk 11]
```diff
@@ -9,6 +9,12 @@ github.com/Shopify/sarama v1.23.0 h1:slvlbm7bxyp7sKQbUwha5BQdZTqurhRoI+zbKorVigQ
...
+github.com/adiweiss/migrator v0.1.2-0.20190811073515-5f6b8b8a442c h1:PpqCVqD0zElK+k2jUfVF9boi5hCyvbv2r9aibLtjyUw=
```
Comment (rhinof) [inline go.sum]: remove references to github.com/adiweiss/migrator

File: go.sum [hunk 12]
```diff
@@ -9,6 +9,12 @@ github.com/Shopify/sarama v1.23.0 h1:slvlbm7bxyp7sKQbUwha5BQdZTqurhRoI+zbKorVigQ
...
+github.com/adiweiss/migrator v0.1.2-0.20190811073515-5f6b8b8a442c/go.mod h1:GKIATFD9JeEji4S4qLNuEIa4X247bQnMZo0iWuKQCpU=
```
Comment (rhinof) [inline go.sum]: remove references to github.com/adiweiss/migrator

File: go.sum [hunk 13]
```diff
@@ -9,6 +9,12 @@ github.com/Shopify/sarama v1.23.0 h1:slvlbm7bxyp7sKQbUwha5BQdZTqurhRoI+zbKorVigQ
...
+github.com/adiweiss/migrator v0.1.2 h1:iEE/PqGWWhv7gKc/OFxeCJSsVIcBLGSKLK+yT3J4Gu0=
```
Comment (rhinof) [inline go.sum]: remove references to github.com/adiweiss/migrator

File: go.sum [hunk 14]
```diff
@@ -9,6 +9,12 @@ github.com/Shopify/sarama v1.23.0 h1:slvlbm7bxyp7sKQbUwha5BQdZTqurhRoI+zbKorVigQ
...
+github.com/adiweiss/migrator v0.1.2/go.mod h1:GKIATFD9JeEji4S4qLNuEIa4X247bQnMZo0iWuKQCpU=
```
Comment (rhinof) [inline go.sum]: remove references to github.com/adiweiss/migrator

File: go.sum [hunk 15]
```diff
@@ -71,8 +77,11 @@ github.com/kr/pty v1.1.1/go.mod h1:pFQYn66WHrOpPYNljwOMqo10TkYh1fy3cYio2l3bCsQ=
 github.com/kr/pty v1.1.8/go.mod h1:O1sed60cT9XZ5uDucP5qwvh+TE3NnUj51EiZO/lmSfw=
 github.com/kr/text v0.1.0 h1:45sCR5RtlFHMR4UwH9sdQ5TC8v0qDQCHnXt+kaKSTVE=
 github.com/kr/text v0.1.0/go.mod h1:4Jbv+DJW3UT/LiOwJeYQe1efqtUx/iVham/4vfdArNI=
+github.com/lib/pq v1.0.0/go.mod h1:5WUZQaWbwv1U+lTReE5YruASi9Al49XbQIvNi/34Woo=
```
Comment (rhinof) [inline go.sum]: where are we using pg ?

Comment (adiweiss) [reply go.sum]: comes with migrator lib

Comment (rhinof) [reply go.sum]: Ok

File: gbus/tx/mysql/migrations.go (hunk 7 above)
Comment (rhinof) [reply gbus/tx/mysql/migrations.go]: comments in go for exported types go by the convention of //MethodName etc...
in this case, the comment should be //TimeoutTableMigration create s the service time out table

File: gbus/tx/mysql/migrations.go (hunk 6 above)
Comment (rhinof) [reply gbus/tx/mysql/migrations.go]:
```suggestion
//OutboxMigrations for creating....
```

File: gbus/tx/mysql/migrations.go (hunk 5 above)
Comment (rhinof) [reply gbus/tx/mysql/migrations.go]: Same as the prev comments

File: gbus/tx/mysql/migrations.go (hunk 7 above)
Comment (rhinof) [reply gbus/tx/mysql/migrations.go]: Otherwise, go lint will produce a warning

File: gbus/tx/mysql/txprovider.go [hunk 16]
```diff
@@ -5,12 +5,11 @@ import (
 	"time"
 
 	_ "github.com/go-sql-driver/mysql" //blank import
-	"github.com/wework/grabbit/gbus"
 	"github.com/wework/grabbit/gbus/tx"
 )
 
 //NewTxProvider returns a new PgProvider
-func NewTxProvider(connStr string) (gbus.TxProvider, error) {
+func NewTxProvider(connStr string) (*tx.Provider, error) {
```
Comment (vladshub) [inline gbus/tx/mysql/txprovider.go]: Why did you switch from interface to an implementation?

Comment (vladshub) [reply gbus/tx/mysql/txprovider.go]: @rhinof you jumped the gun 😉

**Pull request merged**